File: android/src/main/java/com/tencent/bugly/bugly_crash/BuglyCrashPlugin.java
package com.tencent.bugly.bugly_crash;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import com.tencent.bugly.crashreport.CrashReport;
import android.content.Context;
import java.util.Map;
import com.tencent.bugly.crashreport.BuglyLog;
/**
 * Description:bugly oa futter plugin
 * @author rockypzhang
 * @since 2019/5/28
 */
public class BuglyCrashPlugin implements MethodCallHandler {
  /** Plugin registration. */
  private static Context mContext;
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "bugly");
    mContext = registrar.activeContext();
    channel.setMethodCallHandler(new BuglyCrashPlugin());
  }

  /**
   * flutter层调用方法的native层实现
   *
   * @param call flutter方法回调
   * @param result 返回结果
   */
  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("getPlatformVersion")) {
      result.success("Android " + android.os.Build.VERSION.RELEASE);
    } else if (call.method.equals("initCrashReport")){
      String appId = "";
      boolean isDebug = false;
      if (call.hasArgument("appId")) {
        appId = call.argument("appId");
      }
      if (call.hasArgument("isDebug")) {
        isDebug = call.argument("isDebug");
        BuglyCrashPluginLog.isEnable = isDebug;
      }
      CrashReport.initCrashReport(mContext, appId, isDebug);
      BuglyCrashPluginLog.d("onMethodCall initCrashReport");
    } else if (call.method.equals("postException")){
      String type = "";
      String error = "";
      String stackTrace = "";
      Map<String, String> extraInfo = null;
      if (call.hasArgument("type")) {
        type = call.argument("type");
      }
      if (call.hasArgument("error")) {
        error = call.argument("error");
      }
      if (call.hasArgument("stackTrace")) {
        stackTrace = call.argument("stackTrace");
      }
      if (call.hasArgument("extraInfo")) {
        extraInfo = call.argument("extraInfo");
      }
      BuglyCrashPluginLog.d("type:"+type+"error:"+error+" stackTrace:"+stackTrace
        +"extraInfo:"+extraInfo);
      int category = 8;
      CrashReport.postException(category,type,error,stackTrace,extraInfo);
    }else if (call.method.equals("setAppChannel")){
      String appChannel = "";
      if (call.hasArgument("appChannel")) {
        appChannel = call.argument("appChannel");
      }
      CrashReport.setAppChannel(mContext,appChannel);
    }else if (call.method.equals("setAppPackage")){
      String appPackage = "";
      if (call.hasArgument("appPackage")) {
        appPackage = call.argument("appPackage");
      }
      CrashReport.setAppPackage(mContext,appPackage);
    }else if (call.method.equals("setAppVersion")){
      String appVersion = "";
      if (call.hasArgument("appVersion")) {
        appVersion = call.argument("appVersion");
      }
      CrashReport.setAppVersion(mContext,appVersion);
      BuglyCrashPluginLog.d("mContext:"+mContext+" appVersion:"+appVersion);
    }else if (call.method.equals("setUserSceneTag")){
      int userSceneTag = 0;
      if (call.hasArgument("userSceneTag")) {
        userSceneTag = call.argument("userSceneTag");
      }
      CrashReport.setUserSceneTag(mContext,userSceneTag);
      BuglyCrashPluginLog.d("mContext:"+mContext+" appVersion:"+userSceneTag);
    }else if (call.method.equals("setUserId")){
      int userSceneTag = 0;
      String userId = "";
      if (call.hasArgument("userId")) {
        userId = call.argument("userId");
      }
      CrashReport.setUserId(mContext,userId);
      BuglyCrashPluginLog.d("mContext:"+mContext+" appVersion:"+userSceneTag);
    }else if (call.method.equals("putUserData")){
      String userKey = "";
      String userValue = "";
      if (call.hasArgument("userKey")) {
        userKey = call.argument("userKey");
      }
      if (call.hasArgument("userValue")) {
        userValue = call.argument("userValue");
      }
      CrashReport.putUserData(mContext, userKey, userValue);
      BuglyCrashPluginLog.d("userKey:"+userKey+" userValue:"+userValue);
    }else if (call.method.contains("log")){
      buglyLog(call);
    }else if (call.method.contains("setIsDevelopmentDevice")){
      boolean isDevelopmentDevice = false;
      if (call.hasArgument("isDevelopmentDevice")) {
        isDevelopmentDevice = call.argument("isDevelopmentDevice");
      }
      CrashReport.setIsDevelopmentDevice(mContext, isDevelopmentDevice);
      BuglyCrashPluginLog.d("isDevelopmentDevice:"+isDevelopmentDevice);
    } else {
      result.notImplemented();
    }
  }

  /**
   * 打印上报用户自定义日志
   *
   * @param call flutter方法回调
   * @param content 内容
   */
  private void buglyLog(MethodCall call){
    String tag = "";
    String content = "";
    if (call.hasArgument("tag")) {
      tag = call.argument("tag");
    }
    if (call.hasArgument("content")) {
      content = call.argument("content");
    }
    if (call.method.equals("logd")){
      BuglyLog.d(tag,content);
    }else if (call.method.equals("logi")){
      BuglyLog.i(tag,content);
    }else if (call.method.equals("logv")){
      BuglyLog.v(tag,content);
    }else if (call.method.equals("logw")){
      BuglyLog.w(tag,content);
    }else if (call.method.equals("loge")){
      BuglyLog.e(tag,content);
    }
    BuglyCrashPluginLog.d("tag:"+tag+" content:"+content);
  }
}

File: ios/Classes/BuglyCrashPlugin.m
#import "BuglyCrashPlugin.h"
#import <Bugly/Bugly.h>

@implementation BuglyCrashPlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"bugly"
            binaryMessenger:[registrar messenger]];
  BuglyCrashPlugin* instance = [[BuglyCrashPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
  if ([@"getPlatformVersion" isEqualToString:call.method]) {
    result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
  } else if ([@"initCrashReport" isEqualToString:call.method]) {
    NSString *appId = call.arguments[@"appId"];
    NSNumber *debugModeNumber = call.arguments[@"debugMode"];
    NSUInteger *debugModeInt = [debugModeNumber integerValue];
    NSString *serverUrl = call.arguments[@"serverUrl"];
    BuglyConfig *config = [[BuglyConfig alloc] init];
    if(debugModeInt > 0) {
        config.debugMode = YES;
    } else {
        config.debugMode = NO;
    }
    if(![self isEmpty:serverUrl]) {
      config.crashServerUrl = serverUrl;
    }
    config.reportLogLevel = BuglyLogLevelVerbose;
    NSLog(@"initCrashReport appid:%@ debugmode is %d serverUrl is %@",appId,debugModeInt,serverUrl);
    [Bugly startWithAppId:appId config:config];
  } else if ([@"postException" isEqualToString:call.method]) {
     NSLog(@"postException");
     NSString *type = call.arguments[@"type"];
     NSString *error = call.arguments[@"error"];
     NSString *stackTrace = call.arguments[@"stackTrace"];
     NSDictionary *extraInfo = call.arguments[@"extraInfo"];
     NSLog(@"type: %@ , error: %@ , stacktrace: %@",type,error,stackTrace);
     NSArray *stackTraceArray = [stackTrace componentsSeparatedByString:@""];
     [Bugly reportExceptionWithCategory:5 name:error reason:@" " callStack:stackTraceArray extraInfo:extraInfo terminateApp:NO];
  } else if ([@"setAppVersion" isEqualToString:call.method]){
     NSString *appVersion = call.arguments[@"appVersion"];
     NSLog(@"setAppVersion %@",appVersion);
     [Bugly updateAppVersion:appVersion];
  } else if ([@"setUserId" isEqualToString:call.method]){
     NSString *userId = call.arguments[@"userId"];
     NSLog(@"setUserId %@",userId);
     [Bugly setUserIdentifier:userId];
  } else if ([@"putUserData" isEqualToString:call.method]){
     NSString *userKey = call.arguments[@"userKey"];
     NSString *userValue = call.arguments[@"userValue"];
     NSLog(@"putUserData userKey %@ userValue %@ ",userKey,userValue);
     [Bugly setUserValue:userValue forKey:userKey];
  } else if([@"setUserSceneTag" isEqualToString:call.method]) {
     NSNumber *tagNumber = call.arguments[@"userSceneTag"];
     NSUInteger *tagInt = [tagNumber integerValue];
     [Bugly setTag:tagInt];
     NSLog(@"setUserSceneTag");
  } else if ([@"logd" isEqualToString:call.method]){
     NSString *tag = call.arguments[@"tag"];
     NSString *log = call.arguments[@"content"];
     [BuglyLog level:BuglyLogLevelDebug tag:tag log:log];
  } else if ([@"logi" isEqualToString:call.method]){
     NSString *tag = call.arguments[@"tag"];
     NSString *log = call.arguments[@"content"];
     [BuglyLog level:BuglyLogLevelInfo tag:tag log:log];
  } else if ([@"logv" isEqualToString:call.method]){
     NSString *tag = call.arguments[@"tag"];
     NSString *log = call.arguments[@"content"];
     [BuglyLog level:BuglyLogLevelVerbose tag:tag log:log];
  } else if ([@"logw" isEqualToString:call.method]){
     NSString *tag = call.arguments[@"tag"];
     NSString *log = call.arguments[@"content"];
     [BuglyLog level:BuglyLogLevelWarn tag:tag log:log];
  } else if ([@"loge" isEqualToString:call.method]){
     NSString *tag = call.arguments[@"tag"];
     NSString *log = call.arguments[@"content"];
     [BuglyLog level:BuglyLogLevelError tag:tag log:log];
  } else {
    result(FlutterMethodNotImplemented);
  }
}

- (BOOL)isEmpty:(id)object{
    if (object == nil || [object isEqual:[NSNull null]]) {
        return YES;
    } else if ([object isKindOfClass:[NSString class]]) {
        return [object isEqualToString:@""];
    } else if ([object isKindOfClass:[NSNumber class]]) {
        return [object isEqualToNumber:@(0)];
    }
    return NO;
}

@end
